lib/compiler_rt/memcpy.zig

```diff
@@ -8,21 +9,283 @@ comptime {
     }
 }
 
-pub fn memcpy(noalias dest: ?[*]u8, noalias src: ?[*]const u8, len: usize) callconv(.C) ?[*]u8 {
-    @setRuntimeSafety(false);
+const CopyType = if (std.simd.suggestVectorLength(u8)) |vec_size|
+    @Type(.{ .vector = .{
+        .child = u8,
+        .len = vec_size,
+    } })
+else
+    usize;
+
+const copy_alignment = @alignOf(CopyType);
+const copy_size = @sizeOf(CopyType);
+
+comptime {
+    assert(copy_size >= copy_alignment);
+    assert(std.math.isPowerOfTwo(copy_size));
+}
+
+pub const memcpy = if (builtin.mode == .ReleaseSmall)
+    memcpy_small
+else
+    memcpy_fast;
+
+fn memcpy_small(noalias dest: ?[*]u8, noalias src: ?[*]const u8, len: usize) callconv(.C) ?[*]u8 {
```

Comment:
```zig
    @setRuntimeSafety(false);

    for (0..len) |i| {
        dest.?[i] = src.?[i];
    }

    return dest;
```

[generates better code](https://zig.godbolt.org/z/x5vGKq885)

also please follow [style guide](https://ziglang.org/documentation/master/#Style-Guide)

Reply:
this is the -OReleaseSmall version where the goal is to generate fewer bytes, so 2 bytes less code for idiomatic source is a good deal. if you wanted to hit the sweet spot instead you'd just compile compiler_rt with -OReleaseFast and the rest of the application with -OReleaseSmall. there's no niche for this sweet spot you're advocating for.